### PR TITLE
Move REX ssh key management into separate class

### PR DESCRIPTION
**Pull request by wbclark**

### CODE REVIEW

**Review by ekohl — CHANGES_REQUESTED** [review]

File: templates/plugin/remote_execution_ssh.yml.erb [hunk 1]
```diff
@@ -1,9 +1,11 @@
 ---
 :enabled: <%= @module_enabled %>
-:ssh_identity_key_file: <%= scope.lookupvar('::foreman_proxy::plugin::remote_execution::ssh::ssh_identity_path') %>
 :local_working_dir: <%= scope.lookupvar('::foreman_proxy::plugin::remote_execution::ssh::local_working_dir') %>
 :remote_working_dir: <%= scope.lookupvar('::foreman_proxy::plugin::remote_execution::ssh::remote_working_dir') %>
 :kerberos_auth: <%= scope.lookupvar('::foreman_proxy::plugin::remote_execution::ssh::ssh_kerberos_auth') %>
+<% if scope.lookupvar("::foreman_proxy::plugin::remote_execution::ssh::generate_keys -%>
+:ssh_identity_key_file: <%= scope.lookupvar('::foreman_proxy::plugin::remote_execution::ssh::keys::ssh_identity_path') %>
+<% end -%>
```
[the remark this answers is not in the document]
Comment (wbclark) [reply templates/plugin/remote_execution_ssh.yml.erb:8]: fixed syntax...

[the remark this answers is not in the document]
Comment (ekohl) [reply templates/plugin/remote_execution_ssh.yml.erb:8]: It is changing behavior. Prior to this it was up to the user to ensure it was present at the desired location, now it won't show up in the config anymore. I don't think this is really the same thing.

But I'm also not sure what the best solution here is. I think the API of ...::ssh::keys should accept a full path instead and use [`dirname()`](https://forge.puppet.com/modules/puppetlabs/stdlib/reference#dirname) on that path to determine the parent directory. Same for [`basename()`](https://forge.puppet.com/modules/puppetlabs/stdlib/reference#basename) to get the filename.

File: manifests/plugin/remote_execution/ssh/keys.pp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  Stdlib::Absolutepath $ssh_identity_dir = '/var/lib/foreman-proxy/ssh',
+  String $ssh_identity_file = 'id_rsa_foreman_proxy',
+  String $ssh_keygen = '/usr/bin/ssh-keygen',
+) {
+  $ssh_identity_path = "${ssh_identity_dir}/${ssh_identity_file}"
+
+  file { $ssh_identity_dir:
+    ensure => directory,
+    owner  => $foreman_proxy::user,
+    group  => $foreman_proxy::user,
```
Comment (ekohl) [inline manifests/plugin/remote_execution/ssh/keys.pp:27]: I would also make these class parameters. That way you don't need to `include foreman_proxy` to test this.

Comment (wbclark) [reply manifests/plugin/remote_execution/ssh/keys.pp]: How does this look?

File: templates/plugin/remote_execution_ssh.yml.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (wbclark) [reply templates/plugin/remote_execution_ssh.yml.erb]: Done. It's slightly awkward to reconstruct the $ssh_identity_dir when that is already available in the parent scope, but it feels less awkward than any other solution I can think of.

**Comment by wbclark** [comment]
> It does make me wonder: should the name be tied to REX? Just thinking about the future case where Ansible might also need SSH keys but REX is configured to use MQTT.

It's concerning to me too. After staring at it for a bit I don't see a clean way to fix this without breaking some interfaces.

**Comment by wbclark** [comment]
> > It does make me wonder: should the name be tied to REX? Just thinking about the future case where Ansible might also need SSH keys but REX is configured to use MQTT.
> 
> Ansible is REX in our ecosystem. And by that I mean, REX is the umbrella and it has sub-functionality underneath that falls into transports and providers. SSH, http and mqtt being transports, and Scripts and Ansible being providers (or languages) that can be used. It all rolls up to REX though.

This commit cleared things up for me https://github.com/theforeman/smart_proxy_ansible/commit/3f408b6eea4cae58a3fc91c20cfb01c627953fab

Therefore, I think this PR is in a mergeable state unless @ekohl has other feedback?

**Review by ehelms — APPROVED** [review]
@ekohl could you take a final look as well

**Review by ekohl — APPROVED** [review]

**Pull request merged**